### ISSUE TITLE
Optional console output for Windows

### ISSUE THREAD

**Issue by JNechaevsky** [chocolate-doom/chocolate-doom#1166]
AFAIR, we was having some talks on IRC about missing console output for Windows binaries after moving to SDL2. So here it is finally, a blazingly simple approach to make an optional console window. Sorry, I will not provide a PR, but I'll try to explain what should be done.

There are two possible ways.

**1. MSYS compilation with -mconsole flag**

Very simple, by adding `-mconsole` to `CFLAGS` of `configure.ac` file:
```
        WARNINGS="-Wall -Wdeclaration-after-statement -Wredundant-decls"
---     CFLAGS="-O$OPT_LEVEL -g $WARNINGS $orig_CFLAGS"
+++     CFLAGS="-O$OPT_LEVEL -g $WARNINGS $orig_CFLAGS -mconsole"
```

But. This way console window will be non-optional and will appear with any executable, even with setup.exe. I'm _definitely_ not recommending this approach, because this is strictly developer's way, common player win not need console windows always turned on!

**2. Implement an optional console window**

First of all, please note: this will be a Windows-only function, so it should be guarded by `#ifdef _WIN32` macroses. 

First, we need an console invoking function. I believe that `i_system.c` is pretty good place for that, especially because it already having all necessary defs:

```
#ifdef _WIN32
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#else
#include <unistd.h>
#endif
```

Note: `WIN32_LEAN_AND_MEAN` is necessary here, otherwise compilation will bomb out with an error of `boolean` redefinition. Now, the function:

```
#ifdef _WIN32
void I_Windows_Console(void)
{
    // Create a console window
    AllocConsole();

    // Head outputs
    freopen("CONIN$", "r",stdin); 
    freopen("CONOUT$","w",stdout); 
    freopen("CONOUT$","w",stderr);  

    // Set a proper codepage
    SetConsoleOutputCP(CP_UTF8);
    SetConsoleCP(CP_UTF8);
}
#endif
```

Setting a `CP_UTF8` is not really necessary until we will not have a non-US characters. But it's odd: without a UTF8 setting, console window will use a DOS codepage (I don't have an English version of Windows at hand, but in Russian it's a CP866. Thanks Microsoft). That's it, that's all we need for console window. Also, don't forget to add a prototype in i_system.h.

Now, how to invoke it. In my approach, I have placed it into `-devparm` mode, for Chocolate and Crispy this should be different command line parameter of course. Maybe a `-console`? Invoking is even simpler (d_main.c):

```
    boolean		devparm;	// started game with -devparm
+++ boolean		winconsole;	// started game with -console
    boolean         nomonsters;	// checkparm of -nomonsters
    boolean         respawnparm;	// checkparm of -respawn
    boolean         fastparm;	// checkparm of -fast
```
... and then ...
```
//
// D_DoomMain
//
void D_DoomMain (void)
{
    int p;
    char file[256];
    char demolumpname[9];
    int numiwadlumps;

+++     // Create a system console for -console mode. For Windows OS only.
+++ #ifdef _WIN32
+++     winconsole= M_CheckParm ("-console");
+++     if (winconsole)
+++     I_Windows_Console();
+++ #endif 

    I_AtExit(D_Endoom, false);

    // print banner

    I_PrintBanner(PACKAGE_STRING);

```

Finally, the result:

![viola](https://user-images.githubusercontent.com/21193394/58050982-97c3ad00-7b59-11e9-944e-b57d8eca3c05.png)
Please see this commit for the reference: https://github.com/JNechaevsky/russian-doom/commit/db66e3193d4d6dbcd44534347abbffc45301249e

_By the request of @fabiangreffrath (suggestion №108 please)_
_Initial investigation: https://github.com/JNechaevsky/russian-doom/issues/117_

**Comment by fragglet** [comment]
Seems reasonable. I'd recommend putting it in `i_main.c` and leaving it out of the game code entirely.

**Comment by Kroc** [comment]
Would this be able to attach to an existing console -- i.e. when the game is launched from a console, the output pipes into the already existing console. Windows has an API for getting the "current" console.

Could I recommend using `-stdout` instead of `-console` though (or both), as this is already supported by ZDoom-based engines and the consistency would be appreciated (for launchers)

**Comment by fragglet** [comment]
+1 for consistency with other source ports.

Unix has `isatty()` for determining if stdout is already attached to a terminal/console. If there's an equivalent for Windows then we could maybe use this to do what @Kroc suggests.

**Comment by JNechaevsky** [comment]
Agreed, it's a good idea, it will cover all possible cases of running game on Windows and getting console, i.e.:

1. Running from batch file, which is always showing console window.
2. Running from shortcut file, with an ability to add `-stdout` / `-console` cmd line parameter.
3. Running from non-standard file manager like Total Commander, where command line is always available.

Now, some changes must be done in my suggested approach. There are two ways:

1. If we are using cmd line parameter, always show separate console window.
2. If we are **not** using cmd line parameter, use standard windows console. 

By default, outputs are going only in MSYS terminal, cmd.exe is always silent. Ho head outputs to cmd.exe, we have to call `AttachConsole`, just like this...

```
#ifdef _WIN32
void I_Windows_Console(void)
{
    if (-stdout || -console)    // Create a console window
    AllocConsole();
    else                        // Use standard Windows console
    AttachConsole(-1);

    // Head outputs
    freopen("CONIN$", "r",stdin); 
    freopen("CONOUT$","w",stdout); 
    freopen("CONOUT$","w",stderr);  

    // Set a proper codepage
    SetConsoleOutputCP(CP_UTF8);
    SetConsoleCP(CP_UTF8);
}
#endif
```

Note, that outputs heading (`freopen`) is still necessary, otherwise nothing will go to standard Windows console. I have briefly tested `AttachConsole()` with RD and can confirm that it is actually working, but some line spacing is screwed up ([screenshot](https://user-images.githubusercontent.com/21193394/58113440-2b4eb980-7bfe-11e9-8f68-be684c51f5a8.png) of running from cmd.exe). But it seems to be the right track already.

**Comment by JNechaevsky** [comment]
Alright, I'm extremely close to the result we are discussing here: I've got separate console output in case of using `-console` approach, and _cmd.exe_ console output in case of running game from standard Windows console. All the console invoking code is moved into i_main.c.

Everything working fine, except one small thing: when I'm exiting the game with _cmd.exe_ approach, I'm not getting proper return to command prompt until `ENTER` is pressed. It's not a soft-lock, but cursor keep blinking here:

![viola](https://user-images.githubusercontent.com/21193394/58123888-8c35bc00-7c15-11e9-8149-fe57c8564207.png)

Not sure at the moment, what to do.

**Comment by JNechaevsky** [comment]
This seems to be working solution with sending an `ENTER` key, but I'm not sure is it really a solution or just an odd workaround...

In the end of console invoking code, this should be placed:
```
    I_AtExit(W_PressEnter, false);
```

The function itself:
```
void W_PressEnter (void)
{
    keybd_event(VK_RETURN,
                0x0D,
                KEYEVENTF_EXTENDEDKEY,
                0);
}
```

If anyone in the Crew have a better idea, please advice. :/

**Comment by JNechaevsky** [comment]
Here, I've got it working. A bit messy, but at least will show everything in one commit: 
* https://github.com/JNechaevsky/russian-doom/commit/1074146a1c22690c62c7f2d1fd5855cadb5774db

Have a look into i_main.c everything necessary there, the rest of files (*_main.c) is used just for colored titles. It's the best I can do.

**Comment by JNechaevsky** [comment]
Oops, no, sending an `ENTER` key is wrong approach - when the game is started not from cmd.exe and exited, key is still being sent, making possible to start the game (or anything else) again.

**Comment by Kroc** [comment]
Perhaps you need to "detach" from the console when the app shuts down?

**Comment by JNechaevsky** [comment]
That's true. I was trying to invoke `FreeConsole()`, but in fact it's just stops output, not actually detaching it. Another idea was to sending `CTRL+C` breaking, but it's also unsafe - this way it is possible to invoke a system action "copy".

This problem is not rare, there are even some [discussions ](https://stackoverflow.com/questions/1305257/using-attachconsole-user-must-hit-enter-to-get-regular-command-line/3566139) on StackOverflow. Not sure what to do next, though.
